src/main/java/com/mraof/minestuck/computer/editmode/EditmodeLocations.java

```diff
@@ -0,0 +1,371 @@
+	{
+		EditmodeLocations locations = new EditmodeLocations();
+		
+		for(int i = 0; i < locationsTag.size(); i++)
+		{
+			CompoundTag nbt = locationsTag.getCompound(i);
+			
+			ResourceKey<Level> dimension = Level.RESOURCE_KEY_CODEC.parse(NbtOps.INSTANCE, nbt.get("dim")).resultOrPartial(LOGGER::error).orElse(null);
+			
+			int posX = nbt.getInt("x");
+			int posY = nbt.getInt("y");
+			int posZ = nbt.getInt("z");
+			
+			int sourceOrdinal = nbt.getInt("source");
+			
+			locations.addEntry(dimension, new BlockPos(posX, posY, posZ), Source.fromInt(sourceOrdinal));
+		}
+		
+		return locations;
+	}
+	
+	private void checkBlockPosValidation(Player editPlayer, SburbConnection connection, Level editLevel, ResourceKey<Level> editDimension, List<Pair<BlockPos, Source>> allLevelPairs, BlockPos blockIterate)
+	{
+		//TODO locations appear to be added/removed inappropriately
+		
+		//if locations contains the iterated block pos and the entry is no longer valid, remove it. Else if locations did not contain the iterated pos and its valid, add it
+		if(allLevelPairs.stream().anyMatch(blockPosSourcePair -> blockIterate.equals(blockPosSourcePair.getFirst()) && blockPosSourcePair.getSecond() == Source.BLOCK))
+		{
+			if(!isValidBlockSource(connection, editLevel, blockIterate))
+			{
+				connection.removeClientEditmodeLocations(editDimension, blockIterate, Source.BLOCK);
+				
+				int range = MSDimensions.isLandDimension(((ServerPlayer) editPlayer).server, editPlayer.level().dimension()) ? MinestuckConfig.SERVER.landEditRange.get() : MinestuckConfig.SERVER.overworldEditRange.get();
+				
+				//TODO consider adding message indicating what happened
+				if(!canMoveAtPosition(editPlayer, range))
+				{
+					//teleport player to nearest valid location
+					BlockPos nextClosestLocationPos = getClosestPosInDimension(editPlayer);
+					if(nextClosestLocationPos != null)
+					{
+						editPlayer.teleportTo(nextClosestLocationPos.getX() + 0.5D, nextClosestLocationPos.getY() + 1.0D, nextClosestLocationPos.getZ() + 0.5D);
+					}
+				}
+			}
+		} else if(isValidBlockSource(connection, editLevel, blockIterate))
+		{
+			connection.addClientEditmodeLocation(editDimension, blockIterate, Source.BLOCK);
+		}
+	}
+	
+	public static boolean isValidBlockSource(SburbConnection connection, Level level, BlockPos pos)
+	{
+		if(level.getBlockEntity(pos) instanceof ComputerBlockEntity computerBlockEntity)
+		{
+			return !computerBlockEntity.isBroken() && computerBlockEntity.hasProgram(0) && computerBlockEntity.owner.equals(connection.getClientIdentifier());
+		}
+		
+		return false;
+	}
+	
+	/**
+	 * Takes all the flags from canMoveAtPosition() and modifies the players movement/position as needed.
+	 *
+	 * @return Whether the player is out of one or more coordinate bounds
+	 */
+	private boolean handleMovement(Player editPlayer, boolean xInBounds, boolean yInBounds, boolean zInBounds, boolean allMatch, boolean noneMatch)
```

Comment:
One thing that would be a nice safeguard against getting stuck is to instead of setting velocity to 0 along certain axis, we would instead use vector math to remove velocity in the direction away from the closest source.
We could look into it after taking care of some refactor and code cleanup first.
I should have no trouble figuring out how to do it, so feel free to make me do it if you're unsure yourself.

Reply:
That would be appreciated! I wasnt sure how to make movement smoother along the edges